File: backend/roles.py
"""Contribution roles: who may write to live data.

A `roles.yaml` at the ingests root maps a GitHub login to a role. Anyone
authenticated but unlisted is a `contributor` - the safe default - so a new
login can propose edits but never commit to main. The file is editor-edited
(a UI comes later); absent or malformed, everyone is a contributor.

    # ingests/roles.yaml
    markhedleyjones: editor
    somereviewer: reviewer

Roles are ordered contributor < reviewer < editor:
- contributor: propose edits (queued, never committed directly).
- reviewer: + approve/reject proposals; own edits commit directly.
- editor: + manage roles.
"""

from __future__ import annotations

from pathlib import Path

import yaml

ROLES = ("contributor", "reviewer", "editor")
DEFAULT_ROLE = "contributor"


def load_roles(ingests_path: Path) -> dict[str, str]:
    """The login -> role map from `ingests/roles.yaml`. Logins are lowercased;
    entries with an unknown role are dropped. Empty on any read/parse error."""
    path = ingests_path / "roles.yaml"
    if not path.exists():
        return {}
    try:
        data = yaml.safe_load(path.read_text())
    except (OSError, yaml.YAMLError):
        return {}
    if not isinstance(data, dict):
        return {}
    return {
        login.lower(): role
        for login, role in data.items()
        if isinstance(login, str) and role in ROLES
    }


def role_of(login: str | None, ingests_path: Path) -> str:
    """A login's role, defaulting to contributor when unlisted or absent."""
    if not login:
        return DEFAULT_ROLE
    return load_roles(ingests_path).get(login.lower(), DEFAULT_ROLE)


def at_least(role: str, minimum: str) -> bool:
    """True when `role` ranks at or above `minimum` in the hierarchy."""
    rank = ROLES.index(role) if role in ROLES else 0
    floor = ROLES.index(minimum) if minimum in ROLES else 0
    return rank >= floor


def count_editors(roles_map: dict[str, str]) -> int:
    """How many logins hold the editor role - used to refuse the last editor's
    removal so the role file can never lock everyone out."""
    return sum(1 for role in roles_map.values() if role == "editor")


def save_roles(ingests_path: Path, roles_map: dict[str, str]) -> Path:
    """Write the login -> role map to `ingests/roles.yaml`, lowercased and
    key-sorted for a stable diff; entries with an unknown role are dropped.
    Returns the written path (the caller commits it)."""
    clean = {
        login.lower(): role
        for login, role in roles_map.items()
        if isinstance(login, str) and login.strip() and role in ROLES
    }
    # A fixed header keeps the file self-documenting after a UI edit (yaml.dump
    # would otherwise strip any hand-written comment).
    header = (
        "# Contribution roles. Unlisted authenticated users default to `contributor`\n"
        "# (propose-only). Roles: contributor < reviewer < editor.\n"
    )
    body = (
        yaml.safe_dump(clean, default_flow_style=False, sort_keys=True) if clean else ""
    )
    path = ingests_path / "roles.yaml"
    path.write_text(header + body)
    return path

File: edge/lib/roles.ts
/**
 * Contribution roles at the edge - the production write gate.
 *
 * The workbench's Python backend has its own copy of this model
 * (`backend/roles.py`), but production does NOT run that backend: it runs this
 * edge. Until this existed, every write route here gated on "is there a session"
 * alone, so ANY GitHub login could commit to the live ingests repo. This closes
 * that.
 *
 * `roles.yaml` lives at the ingests repo root and maps a GitHub login to a role.
 * Anyone authenticated but unlisted is a `contributor` - the safe default - so a
 * new login can never write. Absent or malformed, EVERYONE is a contributor
 * (fail closed: a missing role file must not grant access).
 *
 *     # ingests/roles.yaml
 *     markhedleyjones: admin
 *     somereviewer: reviewer
 *
 * Roles are ordered contributor < reviewer < editor < admin (cumulative), and
 * must stay in step with backend/roles.py:
 * - contributor: propose only; no writes here.
 * - reviewer: + record writes, curation decisions.
 * - editor: + the higher-impact ops (archive, article directives).
 * - admin: + role management.
 */

export const ROLES = ["contributor", "reviewer", "editor", "admin"] as const;
export type Role = (typeof ROLES)[number];
export const DEFAULT_ROLE: Role = "contributor";

/**
 * Parse `roles.yaml`. Deliberately a small line parser rather than a YAML
 * dependency: the file is a flat `login: role` map with `#` comments, and the
 * gate must not inherit a parser's surprises. Anything that isn't a clean
 * `login: knownRole` line is ignored, so a malformed entry silently grants
 * nothing instead of throwing the gate open.
 */
export function parseRoles(text: string): Record<string, Role> {
  const out: Record<string, Role> = {};
  for (const raw of text.split("\n")) {
    const line = raw.trimEnd();
    if (!line || line.trimStart().startsWith("#")) continue;
    // INDENTED lines belong to some nested structure, not the top-level
    // login->role map. Trimming first would read `nested:\n  role: admin` as a
    // login literally called "role" with admin - so only column-0 keys count.
    if (/^\s/.test(line)) continue;
    const m = line.match(/^([A-Za-z0-9-_]+)\s*:\s*"?([a-z]+)"?\s*$/);
    if (!m) continue;
    const [, login, role] = m;
    if ((ROLES as readonly string[]).includes(role)) {
      out[login.toLowerCase()] = role as Role;
    }
  }
  return out;
}

/** A login's role, defaulting to contributor when unlisted. */
export function roleOf(login: string | undefined, roles: Record<string, Role>): Role {
  if (!login) return DEFAULT_ROLE;
  return roles[login.toLowerCase()] ?? DEFAULT_ROLE;
}

/** True when `role` ranks at or above `minimum`. Unknown roles rank lowest. */
export function atLeast(role: Role, minimum: Role): boolean {
  const rank = ROLES.indexOf(role);
  const floor = ROLES.indexOf(minimum);
  return rank >= 0 && floor >= 0 && rank >= floor;
}
